contracts/infrastructure/zap2/ZapContract2.sol

```diff
@@ -0,0 +1,397 @@
+// SPDX-License-Identifier: ISC
+/**
+* By using this software, you understand, acknowledge and accept that Tetu
+* and/or the underlying software are provided “as is” and “as available”
+* basis and without warranties or representations of any kind either expressed
+* or implied. Any use of this open source software released under the ISC
+* Internet Systems Consortium license is done at your own risk to the fullest
+* extent permissible pursuant to applicable law any and all liability as well
+* as all warranties, including any fitness for a particular purpose with respect
+* to Tetu and/or the underlying software and the use thereof are disclaimed.
+*/
+pragma solidity 0.8.4;
+
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import "../../openzeppelin/IERC20.sol";
+import "../../openzeppelin/SafeERC20.sol";
+import "../../openzeppelin/ReentrancyGuard.sol";
+import "../../base/governance/Controllable.sol";
+import "../../base/interface/ISmartVault.sol";
+import "../../base/interface/IStrategy.sol";
+import "../../base/interface/IController.sol";
+import "../../third_party/uniswap/IUniswapV2Pair.sol";
+import "../../third_party/uniswap/IUniswapV2Router02.sol";
+import "../price/IPriceCalculator.sol";
+import "./IMultiSwap2.sol";
+import "./IZapContract2.sol";
+
+/// @title Dedicated solution for interacting with Tetu vaults.
+///        Able to zap in/out assets to vaults
+/// @author belbix
+contract ZapContract2 is IZapContract2, Controllable, ReentrancyGuard {
+  using SafeMath for uint256;
+  using SafeERC20 for IERC20;
+
+  string public constant VERSION = "2.0.0";
+
+  IMultiSwap2 public multiSwap;
+  mapping(address => uint256) calls;
+  mapping(address => address) public factoryToRouter;
+
+/*  struct ZapInfo {
+    address lp;
+    address tokenIn;
+    address asset0;
+    address[] asset0Route;
+    address asset1;
+    address[] asset1Route;
+    uint256 tokenInAmount;
+    uint256 slippageTolerance;
+  }*/
+
+  constructor(address _controller, address _multiSwap, address[] memory _factories,
+    address[] memory _routers) {
+    initialize(_controller, _multiSwap, _factories, _routers);
+  }
+
+  function initialize(address _controller, address _multiSwap, address[] memory _factories,
+    address[] memory _routers) public initializer {
+    require(_multiSwap != address(0), "ZC: zero multiSwap address");
+    Controllable.initializeControllable(_controller);
+    multiSwap = IMultiSwap2(_multiSwap);
+    for (uint256 i = 0; i < _factories.length; i++) {
+      factoryToRouter[_factories[i]] = _routers[i];
+    }
+  }
+
+  modifier onlyOneCallPerBlock() {
```

Comment:
let's also make protection against contracts too
it will never used with contracts